bin/propolis-server/src/lib/initializer.rs

```diff
@@ -610,74 +617,93 @@ impl<'a> MachineInitializer<'a> {
-                let prev = self.crucible_backends.insert(disk_id, backend);
-                if prev.is_some() {
-                    return Err(MachineInitError::DuplicateCrucibleBackendId(
-                        disk_id,
-                    ));
-                }
+            if let Some(crucible) = crucible {
+                let crucible =
+                    match self.crucible_backends.entry(backend_id.clone()) {
+                        std::collections::btree_map::Entry::Occupied(_) => {
+                            return Err(
+                                MachineInitError::DuplicateCrucibleBackendId(
+                                    backend_id.clone(),
+                                ),
+                            );
+                        }
+                        std::collections::btree_map::Entry::Vacant(e) => {
+                            e.insert(crucible)
+                        }
+                    };
 
-                let Some(block_size) = block_size else {
+                let Some(block_size) = crucible.block_size().await else {
                     slog::error!(
                         self.log,
                         "Could not get Crucible backend block size, \
                         virtual disk metrics can't be reported for it";
-                        "disk_id" => %disk_id,
+                        "disk_id" => %backend_id,
+                    );
+                    continue;
+                };
+
+                let Ok(volume_id) = crucible.get_uuid().await else {
+                    slog::error!(
+                        self.log,
+                        "Could not get Crucible volume ID, \
+                        virtual disk metrics can't be reported for it";
+                        "disk_id" => %backend_id,
                     );
                     continue;
                 };
 
-                // Register the block device as a metric producer, if we've been
-                // setup to do so. Note we currently only do this for the Crucible
-                // backend, in which case we have the disk ID.
                 if let Some(registry) = &self.producer_registry {
                     let stats = VirtualDiskProducer::new(
                         block_size,
                         self.properties.id,
-                        disk_id,
+                        volume_id,
                         &self.properties.metadata,
                     );
+
                     if let Err(e) = registry.register_producer(stats.clone()) {
                         slog::error!(
                             self.log,
                             "Could not register virtual disk producer, \
                             metrics will not be produced";
-                            "disk_id" => %disk_id,
+                            "disk_id" => %backend_id,
+                            "volume_id" => %volume_id,
                             "error" => ?e,
```

Comment:
What's the difference here between `volume_id` and `disk_id`?

Reply:
The distinction here is meant to be between "the ID that was used for the disk element in the spec" and "the ID we got back from `Volume::get_uuid`," but in practice I think these are (currently) always the same, because Nexus constructs VCRs that way (i.e. it always puts the disk ID into the relevant `id` field in the top-level `VolumeConstructionRequest::Volume` that it writes to the DB).